core/src/main/java/org/bitcoinj/store/SPVBlockStore.java

```diff
@@ -131,44 +135,91 @@ public SPVBlockStore(NetworkParameters params, File file, int capacity, boolean
+                // efficient times, which may mean that until the map is deallocated the data on disk is randomly
+                // inconsistent. However the only process accessing it is us, via this mapping, so our own view will
+                // always be correct. Once we establish the mmap the underlying file and channel can go away. Note that
+                // the details of mmapping vary between platforms.
+                buffer = channel.map(FileChannel.MapMode.READ_WRITE, 0, randomAccessFile.length());
+                buffer.get(currentHeader);
             } else {
+                log.info("Creating new SPV block chain file: " + file);
+                randomAccessFile.setLength(fileLength);
+                // Map it into memory read/write. See above comment.
+                buffer = channel.map(FileChannel.MapMode.READ_WRITE, 0, fileLength);
+                initNewStore(params.getGenesisBlock());
+            }
+
+            // Maybe migrate V1 to V2 format.
+            if (Arrays.equals(HEADER_MAGIC_V1, currentHeader)) {
+                long currentLength = randomAccessFile.length();
+                long currentBlocksLength = currentLength - FILE_PROLOGUE_BYTES;
+                if (currentBlocksLength % RECORD_SIZE_V1 != 0)
+                    throw new BlockStoreException(
+                            "File size on disk indicates this is not a V1 block store: " + currentLength);
+                int currentCapacity = (int) (currentBlocksLength / RECORD_SIZE_V1);
+                log.info("Migrating SPV block chain file containing " + currentCapacity + " blocks from V1 to V2 " +
+                        "format: " + file);
+
+                randomAccessFile.setLength(fileLength);
+                // Map it into memory again because of the length change.
+                buffer.force();
+                buffer = channel.map(FileChannel.MapMode.READ_WRITE, 0, fileLength);
+
+                // migrate magic header
+                ((Buffer) buffer).rewind();
+                buffer.put(HEADER_MAGIC_V2);
+
+                // migrate headers
+                final byte[] zeroPadding = new byte[20]; // 32 (V2 work) - 12 (V1 work)
+                for (int i = currentCapacity - 1; i >= 0; i--) {
+                    byte[] record = new byte[RECORD_SIZE_V1];
+                    buffer.position(FILE_PROLOGUE_BYTES + i * RECORD_SIZE_V1);
+                    buffer.get(record);
+                    buffer.position(FILE_PROLOGUE_BYTES + i * RECORD_SIZE_V2);
+                    buffer.put(record, 0, 32); // hash
+                    buffer.put(zeroPadding);
+                    buffer.put(record, 32, RECORD_SIZE_V1 - 32); // work, height, block header
+                }
+
+                // migrate cursor
+                int cursorRecord = (getRingCursor() - FILE_PROLOGUE_BYTES) / RECORD_SIZE_V1;
+                setRingCursor(FILE_PROLOGUE_BYTES + cursorRecord * RECORD_SIZE_V2);
```

Comment:
Could the migration be moved to a static helper method for readability? If this change is merged as is the constructor will be 110 lines of code.

I'm also wondering if the upgrade should be made (effectively) atomic. Obviously this would require more free storage to do, but it would protect for the case of power-failure in the middle of conversion, etc.

Reply:
Extracted helper method via 7383f1d25526c1e96ba331bbb865413359d6a1c5.

Regarding the atomicity, effectively we'd need to create a second file, migrate into that without touching the old file and then rename the new file over the old. We do this with wallet saves. However since migration happens only once and presumibly at app start I would not expect crashing processes etc. Do you think this is a real issue?

Reply:
> Extracted helper method via [7383f1d](https://github.com/bitcoinj/bitcoinj/commit/7383f1d25526c1e96ba331bbb865413359d6a1c5).

Looks good!

Reply:
> However since migration happens only once and presumibly at app start I would not expect crashing processes etc. Do you think this is a real issue?

Well Bitcoin Wallet has a large installed base. If a phone runs out of battery or crashes for some reason during the upgrade, what will happen? How do you recover from a corrupted SPVStore file?

Reply:
> How do you recover from a corrupted SPVStore file?

The blockchain will be reset and the wallet replayed.